Tests/Tests/DDOSLoggingTests.m

```diff
@@ -31,16 +31,16 @@ - (void)tearDown {
 }
 
 - (void)testRegisterLogger {
-    if(@available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOS 3.0, *)) {
-        DDOSLogger *logger = [DDOSLogger sharedInstance];
+    if (@available(iOS 10.0, macOS 10.12, tvOS 10.0, watchOS 3.0, *)) {
+        __auto_type logger = [DDOSLogger new];
         [DDLog addLogger:logger];
-        XCTAssertEqualObjects(logger.loggerName, @"cocoa.lumberjack.osLogger");
-        XCTAssertEqualObjects(logger, DDLog.allLoggers[0]);
+        XCTAssertEqualObjects(logger.loggerName, [DDOSLogger new].loggerName);
```

Comment:
@ffried so, constants are needed?
And yes, it checks that item is exactly OSLogger and nothing else. Checking for unknown string is useless in case of subclassing.

Reply:
And I really don't understand why these tests exist. They don't check anything except adding to DDLog array. So, maybe just remove them?

Reply:
An `OSLogger` is explicitly allocated here. So no subclass will fail this test unless a subclass is explicitly created here, in which case the tests needs to be adjusted anyways.
Since a constant is returned by `loggerName`, we should also test said constant. This will prevent e.g. copy-paste errors.

Reply:
> And I really don't understand why these tests exist. They don't check anything except adding to DDLog array. So, maybe just remove them?

I've made the experience, that in some cases it's better to have some slightly "senseless" tests than no tests at all. As long as they don't cause any problems, they don't hurt anyone. 🙂

Reply:
```objective-c
        XCTAssertEqualObjects(logger.loggerName, [DDOSLogger new].loggerName); // checks that explicit logger has exactly name of OSLogger, no matter how it is acquired. Name is immutable.
        XCTAssertEqualObjects(logger, DDLog.allLoggers.firstObject); // checks that object is added to DDLog shared array.
        XCTAssertEqualObjects(logger.loggerName, @"cocoa.lumberjack.osLogger"); // checks that copy-pasted arbitrary string is equal to current logger name. wtf?
```

Reply:
`loggerName` being immutable (from the outside) does not imply that it does not change. 😉

Reply:
@ffried Checking constants in copy-pasting their values is kind of absurd?

Reply:
> `// checks that copy-pasted arbitrary string is equal to current logger name. wtf?`

Nope. It checks that the logger name is what it should be. Imagine if someone used the library and checked for this constant in code to identify an `OSLogger`. Now if by some accident the name is changed, and this test would not exist, all tests would be green. But the code of said user of the library would suddenly start to fail.
This test makes sure this does not happen. And yes, constants are sometimes duplicated in production code and test code... 🤷‍♂️

Reply:
@ffried Arbitrary string is the worst test case ever. Ok, it is better to create new constants like `kLoggersNames`.

Reply:
> Checking constants in copy-pasting their values is kind of absurd?

How do you secure an important file? You store it in two places.
How do you make sure a constant stays constant? Same procedure. 😉

Reply:
> ok, it is better to create new constants like `kLoggersNames`.

Maybe, yes. But then you would need to check whether said constants contain the correct string and we're back to where we are now. 😉

Reply:
@ffried 
No, in case of `kLoggersNames` there is only one place where all arbitrary data is stored. It is an unique identifier that points to unique resource which is bundled into framework.

Reply:
True. But how do you assert the correctness of said unique identifier?
Again, see above's use case:
> Imagine if someone used the library and checked for this constant in code to identify an OSLogger. Now if by some accident the name is changed, and this test would not exist, all tests would be green. But the code of said user of the library would suddenly start to fail.

Reply:
@ffried 
Bad example with OSLogger. It should have subsystem and subcategory. Nobody will check for loggerName. Everyone will check for class name.
And if I like to access to OSLogger via identifier, I will use constant that library gives me, not an arbitrary string.

Testing uniqueness of identifiers... Well, how do you test that Product bundle identifier is unique? Did you write test for it? Where is a test that check prefix of bundle identifier for each target in project? Where is one test that check .xcconfig?

Reply:
If there is a constant, the user should use said constant, yes. Fair point.

I've made the experience, that tests for constant tend to reduce simple mistakes. If someone **wants** to change the constant, they'll just have to change it in two places - somewhat like a "do you really want to ..."-confirmation.

Uniqueness is hard to test, yes, but correctness (which in this case means string equality) isn't. And again, if you have to adjust the constant **and** a test, simple typo mistakes are less likely made.

(Also, bundle identifier - or generally most build settings - is a bad example as well. I don't write them in my code. Uniqueness of a bundle identifier is something Apple has to check, not me.)

Reply:
Either way, to come back to the point: We should at least remove the `XCTAssertEqualObjects(logger.loggerName, [DOSLogger new].loggerName);` line then. It falsely marks the `loggerName` as tested.

Reply:
@ffried 
Of course, but if this PR revealed new information about constants that should exist in such a complex library, it is better to elaborate them here and merge with correct tests as part of tests.

Constants... I hope that you have ever used Security framework. It restricts usage of arbitrary strings and bind every string to constant name. And the main reason they (Apple) start developing Security is binary compatibility. Because OpenSSL is a big hole of holes. ( even in JWT library in C I saw an example of patching OpenSSL by adding necessary functions in case of new version ).

Again, Constants give user a contract to access items correctly in future. If person uses hardcoded string - it is **his** choice of "simplicity". If person uses smart constant that always grants him correct access to object - it is his choice also. But this choice has support by library.

For example, recent post from Apple:

[Updated Certificate for Game Center Authentication](https://developer.apple.com/news/?id=10312018a) 

> Make sure your app always retrieves and uses the current publicKeyUrl value so it automatically uses the new certificate. If your app caches the certificate or **hardcodes** the certificate URL, it **will require an update**.

Besides, one more thing.
Constants can be public and they intended to give access ( only one way to do it. ). It is public API.

And one more.
Arbitrary strings will fail in **runtime**, not in **compile-time**. That is what Swift and most of developers want.
If you change constant name, it will fail in compile time and user could catch it before release build.

Reply:
Don't get me wrong. Constants are great. And introducing an extensible enum for logger names (that would then map to a RawRepresentable struct in Swift) is definitively a great idea (but IMHO should be its own PR).
My points are the following:
- The assertion like it is now is useless -> should be removed.
- Currently, there aren't any constants, so this "arbitrary string" should be tested. No matter whether it's a good thing that it's arbitrary or not.

I do use RawRepresentable structs or enums for everything in Swift (depending on if they need to be extensible or not). E.g. I never use Strings as storyboard identifiers, I've got a RawRepresentable struct for this. So this not at all about not using constants. I'm all for it.

What I was saying is, that sometimes a constant leaves the scope it is defined in. And in this case it does so as normal string. This means, that even though in our scope, a constant is surely the better solution, it's still a breaking change if the **value** behind the constant changes (not its name!). And to prevent this happening by accident, I sometimes write tests for them.

Reply:
@ffried 
So, this PR discussion evolved into RawRepresentatble struct and constants.
Is there any reason why not to introduce them here?

Reply:
This PR clearly states cleanup. Cleanup IMHO shouldn't introduce new features. Constants would be exactly that. 🙂